File: frontend/lib/prefill.ts
import type { Question } from "./api";

/**
 * Turn extracted document fields into entity-table rows.
 *
 * A large operator will not retype a vendor book that already exists in their
 * purchase orders, so the tables have to be fillable from the documents. The
 * mapping is deliberately conservative: it fills the columns a document can
 * actually evidence and leaves the rest at the pack default, because a
 * confidently wrong lead time is worse than an obviously missing one.
 *
 * Everything produced here lands in an editable table and is labelled as coming
 * from documents, so the operator confirms rather than trusts.
 */

/** Entity column name -> extraction fields that can fill it, best first. */
const SOURCES: Record<string, string[]> = {
  name: ["supplier_name"],
  vendor: ["supplier_name"],
  supplier: ["supplier_name"],
  provider: ["supplier_name"],
  country: ["country"],
  origin: ["country"],
  annual_spend: ["total_value_usd"],
  annual_import_value: ["total_value_usd"],
  annual_cost: ["total_value_usd"],
  lead_time_days: ["lead_time_days"],
  hs_chapter: ["hs_code"],
};

/**
 * Money the operator pays out. This gate is the whole safety of the feature.
 *
 * Every document type the extractor understands is procurement paperwork: a
 * purchase order, an invoice, a customs entry. Those evidence a counterparty
 * you BUY from and nothing else. Matching on column names alone would happily
 * turn a supplier invoice into a wealth-management advisor holding $968,200 of
 * client AUM, or a clinical trial funded by a bearings manufacturer, because
 * "name plus an amount" fits those shapes too.
 *
 * So a table is fillable only if it records what you spend. Revenue-side
 * tables (clients, relationships, AUM, trial value) are never inferred from a
 * purchase order, and the panel says why.
 */
const SPEND_COLUMNS = new Set(["annual_spend", "annual_import_value", "annual_cost"]);

export interface DocRow {
  filename: string;
  doc_type: string;
  confidence: number;
  fields: Record<string, any>;
}

export interface PrefillResult {
  /** question id -> rows we can offer for that table */
  rows: Record<string, Record<string, any>[]>;
  /** question id -> why nothing was produced, when nothing was */
  skipped: Record<string, string>;
  supplierCount: number;
  documentCount: number;
}

function num(v: any): number | null {
  const n = typeof v === "string" ? parseFloat(v.replace(/[^0-9.\-]/g, "")) : v;
  return typeof n === "number" && isFinite(n) ? n : null;
}

/** Group documents by the counterparty they name. One party, one row. */
function groupBySupplier(docs: DocRow[]) {
  const groups = new Map<string, Record<string, any>[]>();
  for (const d of docs) {
    const name = String(d.fields?.supplier_name || "").trim();
    if (!name) continue;
    const key = name.toLowerCase();
    groups.set(key, [...(groups.get(key) || []), d.fields]);
  }
  return groups;
}

export function buildPrefill(questions: Question[], docs: DocRow[]): PrefillResult {
  const entityQs = questions.filter((q) => q.type === "entity_list" && q.fields?.length);
  const groups = groupBySupplier(docs);
  const rows: Record<string, Record<string, any>[]> = {};
  const skipped: Record<string, string> = {};

  for (const q of entityQs) {
    const fields = q.fields || [];
    // A table is fillable only if documents can evidence its identifying column.
    const idField = fields.find((f) => SOURCES[f.name]?.includes("supplier_name"));
    if (!idField) {
      skipped[q.id] = "these documents do not describe this";
      continue;
    }
    if (!fields.some((f) => SPEND_COLUMNS.has(f.name))) {
      skipped[q.id] = "purchase documents evidence what you buy, not this";
      continue;
    }
    if (groups.size === 0) {
      skipped[q.id] = "no counterparty named in these documents";
      continue;
    }

    const built: Record<string, any>[] = [];
    groups.forEach((entries, key) => {
      const row: Record<string, any> = {};
      let evidenced = 0;

      for (const f of fields) {
        const sources = SOURCES[f.name];
        if (!sources) continue;
        for (const src of sources) {
          const values = entries.map((e) => e?.[src]).filter((v) => v !== null && v !== undefined && v !== "");
          if (!values.length) continue;

          if (src === "total_value_usd") {
            // Documents are individual orders, so the book value is their sum,
            // but one shipment usually arrives as several documents: a PO, its
            // invoice and the customs entry all carry the same figure. Summing
            // those would report spend at two or three times the truth, which
            // is the kind of error a buyer spots immediately. Identical values
            // from one counterparty are treated as one transaction.
            const seen = new Set<number>();
            let total = 0;
            for (const v of values.map(num)) {
              if (v === null || v <= 0) continue;
              const key = Math.round(v * 100);
              if (seen.has(key)) continue;
              seen.add(key);
              total += v;
            }
            if (total > 0) { row[f.name] = Math.round(total); evidenced++; }
          } else if (src === "lead_time_days") {
            const ns = values.map(num).filter((n): n is number => n !== null);
            if (ns.length) { row[f.name] = Math.round(Math.max(...ns)); evidenced++; }
          } else if (src === "hs_code") {
            const code = String(values[0]).replace(/\D/g, "").slice(0, 4);
            if (code.length === 4) { row[f.name] = code; evidenced++; }
          } else {
            row[f.name] = String(values[0]); evidenced++;
          }
          break;
        }
      }

      // Name the row after whatever actually identifies it. A supplier name is
      // right for a vendor table and wrong for an import line, where the
      // classification and origin are what distinguish one line from another.
      if (row.hs_chapter) {
        const origin = row.origin || row.country || entries[0]?.country;
        row[idField.name] = origin ? `HS ${row.hs_chapter} from ${origin}` : `HS ${row.hs_chapter}`;
      } else if (!row[idField.name]) {
        row[idField.name] = entries[0]?.supplier_name || key;
      }
      // One column alone is a name with no numbers behind it, which is noise.
      if (evidenced >= 2) built.push(row);
    });

    if (built.length) rows[q.id] = built;
    else skipped[q.id] = "documents named a counterparty but carried no usable figures";
  }

  return {
    rows,
    skipped,
    supplierCount: groups.size,
    documentCount: docs.length,
  };
}

/**
 * Fill a table's blanks so no row is half-empty, without inventing risk.
 *
 * The rule that matters is the boolean one. Copying the first sample row would
 * mark every extracted vendor sole-source, because the pack's first example
 * vendor happens to be, and nothing in a purchase order says otherwise. That
 * one inherited flag roughly doubles derived sole-source exposure and inflates
 * the operator's number on evidence that does not exist. An unevidenced flag
 * is therefore false, and the operator ticks the ones that are true.
 *
 * Numbers fall back to the median of the pack's own examples rather than
 * whichever row happened to be first, so an unevidenced column lands on a
 * typical value instead of an arbitrary one.
 */
export function withDefaults(
  row: Record<string, any>,
  fields: { name: string; type: string }[],
  samples: Record<string, any>[] | undefined
): Record<string, any> {
  const out: Record<string, any> = {};
  const rows = samples || [];
  for (const f of fields) {
    if (row[f.name] !== undefined) {
      out[f.name] = row[f.name];
      continue;
    }
    if (f.type === "bool") {
      out[f.name] = false;
      continue;
    }
    const vals = rows.map((s) => s?.[f.name]).filter((v) => v !== undefined && v !== null);
    if (!vals.length) {
      out[f.name] = f.type === "text" ? "" : 0;
      continue;
    }
    if (f.type === "text" || f.type === "choice") {
      out[f.name] = vals[0];
    } else {
      const ns = vals.map(num).filter((n): n is number => n !== null).sort((a, b) => a - b);
      out[f.name] = ns.length ? ns[Math.floor(ns.length / 2)] : 0;
    }
  }
  return out;
}

File: backend/documents/prefill.py
"""Turn extracted documents into rows for the industry pack's own tables.

This runs on the server because the pack lives here: the tables, their columns
and their published defaults are all one import away, and the alternative was a
pile of column-name guessing in the browser.

Three rules decide whether a row is honest, and all three came from watching it
get them wrong:

  One shipment, several documents. A purchase order, its invoice and the
  customs entry all carry the same figure. Summing them reported a $1.18M order
  as $2.37M of spend, so identical values from one entity count once.

  Stocks are not flows. Two custodial statements for one client are that
  client's money counted twice if you add them, and the same is true of record
  counts and annual contract values. Those take the latest or largest value
  rather than a sum; only genuinely transactional amounts sum.

  Never invent a risk flag. Filling blanks from the pack's first example row
  marked every extracted vendor sole-source, because the first example vendor
  is, which roughly doubled derived exposure on no evidence at all. A column no
  document speaks to falls back to the pack's own published default, and is
  reported as unevidenced so the operator checks it.
"""
from __future__ import annotations

from collections import Counter

from .profiles import ANY, FIRST, MAX, SUM, get_profile


def _num(x):
    if isinstance(x, bool) or x is None:
        return None
    try:
        v = float(str(x).replace(",", "").replace("$", "").strip())
    except (TypeError, ValueError):
        return None
    return v if v == v else None  # drop NaN


def _truthy(x) -> bool:
    if isinstance(x, bool):
        return x
    return str(x).strip().lower() in {"true", "yes", "y", "1"}


def _combine(values: list, how: str):
    """Fold the values several documents gave for one column."""
    if how == ANY:
        return any(_truthy(v) for v in values if v is not None)
    clean = [v for v in values if v is not None and v != ""]
    if not clean:
        return None
    if how == FIRST:
        return str(clean[0])
    nums = [n for n in (_num(v) for v in clean) if n is not None and n > 0]
    if not nums:
        return None
    if how == MAX:
        return max(nums)
    # SUM, with identical values treated as one transaction described twice.
    seen, total = set(), 0.0
    for n in nums:
        k = round(n, 2)
        if k in seen:
            continue
        seen.add(k)
        total += n
    return total


def _median(vals: list[float]) -> float:
    ordered = sorted(vals)
    return ordered[len(ordered) // 2]


def _pack_default(samples: list[dict], column: str, kind: str, scale: float = 1.0):
    """The pack's own published stance on a column no document evidenced.

    Booleans take the majority of the pack's examples rather than False, because
    forcing False is only neutral for flags that raise risk. `has_fallback` and
    `has_successor` lower it, so defaulting those to False would quietly make
    every uploaded book look worse than the pack claims a typical one is.

    Counts are scaled to the size of the row they land in. A raw pack median put
    240 subjects enrolled into a trial with a target of 210, which is not merely
    wrong but impossible, and it read as zero enrollment risk on the trial that
    had the least evidence behind it. Scaling by the row's own evidenced size
    keeps an unevidenced count in proportion to the rest of its row.
    """
    vals = [s.get(column) for s in samples if s.get(column) is not None]
    # Free text is left blank, never borrowed. The pack's example sponsor is
    # "Meridian Bio", and copying that into an operator's real trials put an
    # invented company name in their book and collapsed sponsor concentration
    # to a single counterparty. A categorical column is different: its default
    # is one of a fixed set the pack publishes, not a proper noun.
    if kind == "text":
        return ""
    if kind == "bool":
        return Counter([v for v in vals if isinstance(v, bool)]).most_common(1)[0][0] if any(
            isinstance(v, bool) for v in vals) else False
    if not vals:
        return "" if kind == "choice" else 0
    if isinstance(vals[0], (int, float)):
        base = _median([float(v) for v in vals])
        # Shares and rates are already dimensionless, so they must not scale.
        if 0 < base <= 1 and all(0 <= float(v) <= 1 for v in vals):
            return base
        out = base * scale
        return round(out) if float(base).is_integer() else out
    return vals[0]


def _row_scale(row: dict, samples: list[dict], columns: dict) -> float:
    """How big this row is against the pack's typical one.

    Uses the largest evidenced numeric column the pack also publishes, which in
    practice is the money column: contract value, spend, AUM. Falls back to 1.0
    when nothing comparable was evidenced.
    """
    best = 1.0
    for column in columns:
        v = _num(row.get(column))
        if v is None or v <= 1:
            continue
        vals = [float(s[column]) for s in samples
                if isinstance(s.get(column), (int, float)) and not isinstance(s.get(column), bool)]
        vals = [x for x in vals if x > 1]
        if not vals:
            continue
        base = _median(vals)
        if base > 0:
            return max(0.1, min(10.0, v / base))
    return best


def build_prefill(industry: str, documents: list[dict], pack) -> dict:
    """Rows for each entity table the uploaded documents can actually evidence.

    Returns {question_id: {rows, unevidenced, label}} plus a note per table the
    documents could not speak to, so the panel can say why rather than silently
    showing nothing.
    """
    profile = get_profile(industry)
    # Every row of every table in every document. A sponsor revenue schedule is
    # six sponsors, not one, and reading only the first was why real operating
    # reports produced nothing at all.
    fields: list[dict] = []
    for d in documents:
        recs = d.get("records")
        fields.extend(r for r in recs if isinstance(r, dict)) if recs else fields.append(d.get("fields") or {})
    questions = {q.id: q for q in pack.questions if q.type == "entity_list"}

    filled: dict[str, dict] = {}
    skipped: dict[str, str] = {}

    for table in profile.tables:
        q = questions.get(table.question)
        if q is None:
            continue

        # Reports group the same business differently: one lists studies, another
        # lists the sponsors paying for them. Try the table's own key first and
        # fall back, so a schedule that never names a study still populates.
        groups: dict[str, list[dict]] = {}
        for key_field in (table.key, *table.alt_keys):
            for f in fields:
                key = f.get(key_field)
                if key is None or str(key).strip() == "":
                    continue
                groups.setdefault(str(key).strip(), []).append(f)
            if groups:
                break

        if not groups:
            skipped[table.question] = "nothing in these documents identifies one of these"
            continue

        samples = q.default if isinstance(q.default, list) else []
        schema = {f["name"]: f for f in (q.fields or [])}
        rows, unevidenced_cols = [], set()

        for key, entries in groups.items():
            # A row has to be evidenced by more than its own name.
            if not any(entries_have(entries, n) for n in table.needs):
                continue

            row, seen_cols = {}, set()
            for column, (source, how) in table.columns.items():
                if column not in schema:
                    continue
                value = _combine([e.get(source) for e in entries], how)
                if value is None or (how == ANY and value is False
                                     and not entries_have(entries, source)):
                    continue
                row[column] = value
                seen_cols.add(column)

            # Name the row when the key is not itself a name.
            name_col = next((c for c in ("name", "part") if c in schema), None)
            if name_col and not row.get(name_col):
                parts = [str(entries[0].get(f)) for f in table.label_from
                         if entries[0].get(f)]
                row[name_col] = " from ".join(parts) if parts else key
                seen_cols.add(name_col)

            scale = _row_scale(row, samples, table.columns)
            for f in q.fields or []:
                if f["name"] not in row:
                    row[f["name"]] = _pack_default(samples, f["name"], f["type"], scale)
                    unevidenced_cols.add(f["label"])
            rows.append(row)

        if rows:
            filled[table.question] = {
                "label": q.label,
                "rows": rows,
                "unevidenced": sorted(unevidenced_cols),
            }
        else:
            skipped[table.question] = "named in these documents but with no usable figures"

    for qid, q in questions.items():
        if qid not in filled and qid not in skipped:
            skipped[qid] = "these documents do not describe this"

    return {
        "prefill": filled,
        "skipped": skipped,
        "profile": profile.role,
        "note": (
            "Extraction is best-effort. Columns no document evidenced keep this "
            "industry's published default and are listed so you can check them."
        ),
    }


def entries_have(entries: list[dict], field: str) -> bool:
    return any(e.get(field) not in (None, "") for e in entries)
